File: iOS5bookExamples/ch12p302NotificationLeaker/p229NotificationLeaker/FlipsideViewController.m


#import "FlipsideViewController.h"

/* illustrating the rather tricky memory management associated with 
 addObserverForName:...block: under ARC.
 Run the app and flip and flip back.
 If dealloc is never called, you're leaking.
 */

#define which 1 // and try 2 for one answer, and 3 for another

// however, if all you have is individual observers kept in individual ivars,
// the easiest solution is to declare your observer ivar __weak
// this works because the notification center retains the observer until it is removed

// yet another solution, if you *do* have a collection of observers, is to wrap each one
// using NSValue valueWithNonretainedObject: (and extract with nonretainedObjectValue later)

@implementation FlipsideViewController {
    id observer;
}

@synthesize delegate = _delegate;

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    switch (which) {
        case 1: case 2: {
            observer = [[NSNotificationCenter defaultCenter] addObserverForName:@"woohoo" 
                                                                         object:nil queue:nil 
                                                                     usingBlock:^(NSNotification *note) 
            {
                [self description]; // the mere mention of self, explicit or implicit, potentially causes the leak
            }];

            break;
        }
        case 3: {
            __weak FlipsideViewController* wself = self;
            observer = [[NSNotificationCenter defaultCenter] addObserverForName:@"woohoo" 
                                                                         object:nil queue:nil 
                                                                     usingBlock:^(NSNotification *note) 
            {
                FlipsideViewController* sself = wself;
                if (sself) {
                    [sself description];
                }
            }];
            
            break;
        }
    }
}

- (IBAction)done:(id)sender
{
    NSLog(@"observer %@", observer); // prove that observer is valid
    [[NSNotificationCenter defaultCenter] removeObserver:observer];
    switch (which) {
        case 1: case 3: break;
        case 2: observer = nil;
    }
    [self.delegate flipsideViewControllerDidFinish:self];
}

- (void) dealloc {
    NSLog(@"dealloc");
}

@end

File: iOS8bookExamples/bk1ch12p497notificationLeaker/ch12p325NotificationLeaker/FlipsideViewController.swift


import UIKit

protocol FlipsideViewControllerDelegate : class {
    func flipsideViewControllerDidFinish(controller:FlipsideViewController)
}

class FlipsideViewController: UIViewController {
    
    weak var delegate : FlipsideViewControllerDelegate! = nil
    
    var observer : AnyObject! = nil
    
    let which = 0
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        switch which {
        case 0:
            self.observer = NSNotificationCenter.defaultCenter().addObserverForName(
                "woohoo", object:nil, queue:nil) {
                    _ in
                    self.description; return // leak me, leak me
            }
        case 1:
            self.observer = NSNotificationCenter.defaultCenter().addObserverForName(
                "woohoo", object:nil, queue:nil) {
                    [unowned self] _ in // ha ha, fixed it
                    self.description; return
            }
        default:break
        }
    }
    
    override func viewDidDisappear(animated: Bool) {
        super.viewDidDisappear(animated)
        println("unregister")
        NSNotificationCenter.defaultCenter().removeObserver(self.observer)
    }
    
    @IBAction func done (sender:AnyObject!) {
        println("done")
        self.delegate?.flipsideViewControllerDidFinish(self)
    }
    
    // if deinit is not called when you tap Done, we are leaking
    deinit {
        println("deinit")
    }
    
}

extension FlipsideViewController : UIBarPositioningDelegate {
    func positionForBar(bar: UIBarPositioning) -> UIBarPosition {
        return .TopAttached
    }
}
